crates/bevy_ecs/src/system/commands.rs

```diff
@@ -4,40 +4,56 @@ use crate::{
     entity::{Entities, Entity},
     world::World,
 };
-use bevy_utils::tracing::debug;
+use bevy_utils::{anystack::AnyStack, tracing::debug};
 use std::marker::PhantomData;
 
 /// A [`World`] mutation.
 pub trait Command: Send + Sync + 'static {
-    fn write(self: Box<Self>, world: &mut World);
+    fn write(self, world: &mut World);
+}
+
+/// # Safety
+///
+/// This function is only used when the provided `world` is a `&mut World` and
+/// is only ever called once, so it's safe to consume `command`.
+unsafe fn write_command_on_mut_world<T: Command>(command: *mut u8, world: *mut u8) {
+    let world = &mut *world.cast::<World>();
+    let command = command.cast::<T>().read();
+    command.write(world);
 }
 
 /// A queue of [`Command`]s.
 #[derive(Default)]
 pub struct CommandQueue {
-    commands: Vec<Box<dyn Command>>,
+    commands: AnyStack,
 }
 
 impl CommandQueue {
     /// Execute the queued [`Command`]s in the world.
     /// This clears the queue.
     pub fn apply(&mut self, world: &mut World) {
         world.flush();
-        for command in self.commands.drain(..) {
-            command.write(world);
+        // SAFE:
+        // `apply`: the provided function `write_command_on_mut_world` safely
+        // handle the provided [`World`], drops the associate `Command`, and
+        // clears the inner [`AnyStack`].
+        //
+        // `clear`: is safely used because the call to `apply` above
```

Comment:
Not dropping isn't a safety concern.

Reply:
Yes it is.
If you don't drop something like a `Box<T>` that's a memory leak.

Reply:
Ugh I always forget that in rust that not dropping (and leaking memory) isn't technically unsafe. :/